### PR TITLE
[Merged by Bors] - feat(AlgebraicGeometry/EllipticCurve/Projective): implement equations and nonsingularity for projective coordinates

### PR DESCRIPTION
**Pull request by Multramate**
Define a projective point representative over a ring `R` as the type `Fin 3 -> R`, and its equivalence class `PointClass` as a quotient by the usual scaling relation. Define the analogous `equation` and `nonsingular` predicates on `Fin 3 -> F` over a field `F`, and lift `nonsingular` to `PointClass`. This also has minimal API (e.g. it's missing many of the `equation` lemmas) as most computations should ideally be done using the affine API.

This is the first in a series of four PRs leading to #8485 

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/AlgebraicGeometry/EllipticCurve/Projective.lean
Comment (jcommelin) [inline Mathlib/AlgebraicGeometry/EllipticCurve/Projective.lean:100]: Have you considered using https://leanprover-community.github.io/mathlib4_docs/Mathlib/LinearAlgebra/Projectivization/Basic.html#Projectivization ? It could be applied to the vector space `Fin 3 -> R`.
It assumes that the base ring is a division ring, instead of a commutative ring. But that could probably be generalized.
Duplicating that construction will probably lead to a lot of API duplication down the road.

Comment (Multramate) [reply Mathlib/AlgebraicGeometry/EllipticCurve/Projective.lean:100]: Yes, that was the very first thing I considered, but there are a couple of annoyances at the top of my head:
- `Projectivization` has an extra layer of `v ≠ 0` for every vector `v` in `V`, which is not necessary here because it's already implied by `nonsingular`, and I prefer not to have that extra layer. I don't think anyone is currently developing API for concrete nonsingular varieties defined in terms of formal multivariate derivatives.
- As you mentioned, we need this for more than just division rings, and for an arbitrary commutative ring the correct definition is [slightly subtler](https://mathoverflow.net/questions/46116/a-valued-points-of-projective-space), as discussed in the elliptic curves topic in Zulip. We only need these for rings with trivial Picard group so I'm not sure how to incorporate this general restriction into the `Projectivization` file, so currently it's done ad-hoc.
- We don't have `WeightedProjectivization`, so the analogous construction would need to be made for `Jacobian`. If this is done down the line, and everything else works out, I'll be happy to switch over.

I imagine that there should not be too much API duplication, in the sense that the current duplicated code (99 -- 107 in this file) should be most of the duplicate API ever written for our purposes.

File: Mathlib/AlgebraicGeometry/EllipticCurve/Projective.lean
Comment (MichaelStollBayreuth) [inline Mathlib/AlgebraicGeometry/EllipticCurve/Projective.lean:81]: Why are you using `0`, `1`, `2` on the left and `x`, `y`, `z` on the right?
```suggestion
    (u • P) x = u * P x ∧ (u • P) y = u * P y ∧ (u • P) z = u * P z :=
```

File: Mathlib/AlgebraicGeometry/EllipticCurve/Projective.lean
Comment (MichaelStollBayreuth) [inline Mathlib/AlgebraicGeometry/EllipticCurve/Projective.lean:206]: Do you want `≠ 0` here (non-singular on the generic fiber) or `IsUnit` (non-singular in all fibers)?
It probably depends on how this definition will be used. `≠ 0` is certainly easier to work with (and also the correct version in some cases, e.g., when talking about integral points on elliptic curves over the rationals).

File: Mathlib/AlgebraicGeometry/EllipticCurve/Projective.lean
Comment (MichaelStollBayreuth) [inline Mathlib/AlgebraicGeometry/EllipticCurve/Projective.lean:221]: This could be written in terns of `IsUnit Y` and then would not require `[NoZeroDivisors R]`.

File: Mathlib/AlgebraicGeometry/EllipticCurve/Projective.lean
Comment (MichaelStollBayreuth) [inline Mathlib/AlgebraicGeometry/EllipticCurve/Projective.lean:281]: Again, maybe better to require `IsUnit (P z)` and use the inverse; then one can weaken the assumptions on `F`.
Of course, for more general rings, it is possible that the last coordinate is neither a unit nor zero, so not all cases are covered, but this will be the interesting case for example when implementing the ECM factorization method. So I think it makes sense to keep that in mind.

File: Mathlib/AlgebraicGeometry/EllipticCurve/Projective.lean
Comment (Multramate) [reply Mathlib/AlgebraicGeometry/EllipticCurve/Projective.lean:206]: Yes, thank you! All the suggestions here are very valid and I'll try to apply them tomorrow.

Comment (alreadydone) [reply Mathlib/AlgebraicGeometry/EllipticCurve/Projective.lean:206]: Do you want to say one of the derivatives is a unit, or the ideal generated by the derivatives is the unit ideal? I feel the latter would generalize more broadly, e.g. if you have an elliptic curve over the integers (localized away from bad primes) then I think only the latter defines the correct version of nonsingularity. Of course over a local ring (in particular, a valuation ring) both are equivalent. I've previously [observed](https://leanprover.zulipchat.com/#narrow/stream/116395-maths/topic/thoughts.20on.20elliptic.20curves/near/384388729) that if you define the projective space P^n(R) as (n+1)-tuples that generate the unit ideal ⟨1⟩ quotiented by Rˣ, then `P^n(R) → P^n(Frac(R))` is a bijective if R is a Bezout domain (in particular, a PID or a valuation ring). Restricting to the tuples that generates ⟨1⟩ will be necessary to define the reduction of a point, but it could also be covered by the nonsingularity condition if we adopt the latter definition, which would have an advantage similar to [getting rid of the `v ≠ 0` layer](https://github.com/leanprover-community/mathlib4/pull/9416#discussion_r1443732520) in `Projectivization`.

Regarding the comment on integral points: my impression is that they fit more into the affine API rather than the projective API here? I'd also like to learn more bout why `≠ 0` is the correct version in that case ...

Comment (Multramate) [reply Mathlib/AlgebraicGeometry/EllipticCurve/Projective.lean:206]: Right - I think we want that the ideal generated by the partial derivatives is a unit. The current affine definition for nonsingularity is also using `≠ 0`, and I don't believe that that's equivalent to any of the suggestions here. For instance, I don't know how to prove `nonsingular_some` if I were to change the `nonsingular` definition.

I propose that we stick to treating `R` as if it were a field or a local ring just for this PR (and the one for Jacobian coordinates), and use `≠ 0`. We can then correct the definitions all at once in a future PR.

Comment (alreadydone) [reply Mathlib/AlgebraicGeometry/EllipticCurve/Projective.lean:206]: I forgot - did you try to replace the affine nonsingular condition by the condition that the ideal generated by the two partial derivatives is the unit ideal, and did you encounter difficulties? That condition is equivalent to requiring all three partial derivatives generate the unit ideal, because the third derivative $F_Z(x,y,1)$ is equal to $-(xF_X(x,y,1) + yF_Y(x,y,1))$ for $[x:y:1]$ on the curve $F=0$, because of the identity $XF_X(X,Y,Z)+YF_Y(X,Y,Z)+ZF_Z(X,Y,Z)=3F(X,Y,Z)$ for a homogeneous polynomial $F$ of degree 3. (Did I post this somewhere?)

I'd like to get the definition right because to define the reduction homomorphism I do need to restrict to the tuples that generate the unit ideal. It would be enough to require one of the coordinates to be a unit to first construct the homomorphism in the case of valuation rings, but I think it would be convenient to provide the homomorphism for a more general class of rings (probably Bezout domains). It's also nice that the "generates the unit ideal" condition is "functorial" in the sense that it's preserved by arbitrary RingHoms.

It's nice that you implemented group operations in terms of projective coordinates and connected it to the affine operations in the case of a field, but with the right definition of nonsingularity, we can probably show the points form a group under the operations for a much more general class of rings (domains?). But maybe the field case already justifies merging this PR as is; I'll let other reviewers decide ...

Comment (Multramate) [reply Mathlib/AlgebraicGeometry/EllipticCurve/Projective.lean:206]: Yes, you posted this in the original PR. I don't think I would have any difficulties changing the affine nonsingular condition first, and then the projective and Jacobian versions, but like I said I'd like to do it in a separate PR. And yes, by changing the definition of nonsingularity we should be able to prove the group law for these domains.

Comment (alreadydone) [reply Mathlib/AlgebraicGeometry/EllipticCurve/Projective.lean:206]: @jcommelin asked about this PR today, so let's decide now: you think it's easier to merge this PR (and the Jacobian counterpart) first, and later open a PR that changes the definition of `nonsingular` for affine, projective and Jacobian coordinates all at once? I would think it's easier to open a PR to change the affine definition, and then update the projective and Jacobian PRs after it's merged, but if you prefer that way, let's merge this now.

File: Mathlib/AlgebraicGeometry/EllipticCurve/Projective.lean
Comment (Multramate) [reply Mathlib/AlgebraicGeometry/EllipticCurve/Projective.lean:222]: I would say we don't worry about this for now, and focus on the generalisation all in the new PR.

File: Mathlib/AlgebraicGeometry/EllipticCurve/Projective.lean
Comment (Multramate) [reply Mathlib/AlgebraicGeometry/EllipticCurve/Projective.lean:206]: Yes, I would merging this and the Jacobian counterpart now, and do the generalisation later. One reason being that I won't have time to do the new PR until April, but I would prefer to have this ready as soon as possible.

File: Mathlib/AlgebraicGeometry/EllipticCurve/Projective.lean
Comment (alreadydone) [inline Mathlib/AlgebraicGeometry/EllipticCurve/Projective.lean:140]:
```suggestion
def Equation (P : Fin 3 → R) : Prop :=
```

File: Mathlib/AlgebraicGeometry/EllipticCurve/Projective.lean
Comment (alreadydone) [inline Mathlib/AlgebraicGeometry/EllipticCurve/Projective.lean:220]:
```suggestion
def Nonsingular (P : Fin 3 → R) : Prop :=
```

**Review by alreadydone — CHANGES_REQUESTED** [review]
I just realized capitalization are wrong for the defs `nonsingular` and `equation`. Can you correct the Jacobian PR as well? And maybe the affine defs in another PR. Thanks!

**Comment by alreadydone** [comment]
Thanks!
maintainer merge

**Comment by riccardobrasca** [comment]
There is a conflict, but otherwise LGTM.

bors d+

**Comment by riccardobrasca** [comment]
bors merge

**Pull request closed without merging**
